audio/firmware.c

```diff
@@ -1,201 +1,130 @@
- *
- * This program is distributed in the hope that it will be useful,
- * but WITHOUT ANY WARRANTY; without even the implied warranty of
- * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
- * GNU Lesser General Public License for more details.
- *
- * You should have received a copy of the GNU Lesser General Public License
- * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+// Copyright (C) 2017 Kyle Robbertze <krobbertze@gmail.com>
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 2 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+/** \file firmware.c
+ * Initialises the USB audio firmware and handles interrupts
  */
+
+#ifdef DEBUG
 #include <stdio.h>
+#include "debug.h"
+#else
+#define printf(...)
+#define usart_init()
+#endif
 
-#include <fx2regs.h>
-#include <fx2macros.h>
-#include <serial.h>
-#include <delay.h>
 #include <autovector.h>
-#include <setupdat.h>
+#include <delay.h>
 #include <eputils.h>
+#include <fx2macros.h>
+#include <fx2regs.h>
+#include <fx2types.h>
+#include <setupdat.h>
 
 #include "fx2lights.h"
+#include "audiodata.h"
 
 #define SYNCDELAY SYNCDELAY4
-#define REARMVAL 0x80
-#define REARM() EP2BCL=REARMVAL
 
-volatile WORD bytes;
-volatile __bit gotbuf;
-volatile BYTE icount;
 volatile __bit got_sud;
-DWORD lcount;
-__bit on;
+extern BYTE alt_setting;
 
 void main() {
-    REVCTL=0; // not using advanced endpoint controls
+    d1on();
+    d2on();
+    /* Not using advanced endpoint controls */
+    REVCTL=0;
 
-    d1off();
-    on=0;
-    lcount=0;
-    got_sud=FALSE;
-    icount=0;
-    gotbuf=FALSE;
-    bytes=0;
+    got_sud = FALSE;
 
-    // renumerate
-    RENUMERATE_UNCOND(); 
+    /* renumerate */
+    RENUMERATE_UNCOND();
 
     SETCPUFREQ(CLK_48M);
     SETIF48MHZ();
-    sio0_init(57600);
+    usart_init();
 
-    USE_USB_INTS(); 
+    USE_USB_INTS();
     ENABLE_SUDAV();
-    ENABLE_SOF();
-    ENABLE_HISPEED();
     ENABLE_USBRESET();
+    ENABLE_HISPEED();
+    d1off();
 
-    // only valid endpoints is 2
-    EP2CFG = 0xA2; // 10100010
-    SYNCDELAY;
-    EP1OUTCFG &= ~bmVALID;
-    SYNCDELAY;
-    EP4CFG &= ~bmVALID;
-    SYNCDELAY;
-
-    // arm ep2
-    EP2BCL = 0x80; // write once
-    SYNCDELAY;
-    EP2BCL = 0x80; // do it again
-
-    // make it so we enumerate
-    EA=1; // global interrupt enable 
-    printf ( "Done initializing stuff\n" );
-
+    /**
```

Comment:
You have used this construct multiple times, we should add it to fx2lib?

Reply:
Probably a macro for `DISABLEEP` which sets the config for that to `0x7F` would help